### DISCUSSION TITLE
How to get print margin for PDF?

### DISCUSSION THREAD

**Discussion by devsetgo** [py-pdf/pypdf#2701] [General]
I am struggling a bit to understand how I can get the print margin. I have found mediabox, cropbox, etc.., but that appears to show the page size (e.g. 8"x11"). The purpose is to ensure that there is space for a barcode that gets added when processing during the printing process. We occasionally get very small margin sizes, so trying to figure this out as part of a preparation of the documents for printing.

Any help or pointers is appreciated.

**Comment by stefan6419846** [comment]
I recommend you to have a look at the definition of the different boxes, for example inside the PDF 2.0 specification in section 14.11.2.1 and the corresponding figure 116. Their exact meaning always depends on the use case:

> How the various boundaries are used depends on the purpose to which the page is being put.

**Reply by devsetgo** [reply]
Thank you for the quick reply. I was reading through the documentation and get somewhat where my gap is. What I think my gap is the "Page Size"  [sample below](https://wiki.scribus.net/canvas/PDF_Boxes_:_mediabox,_cropbox,_bleedbox,_trimbox,_artbox). I assume it would be getting the difference between page size and mediabox for example to calculate the margin. What I cannot figure out is how to get that page size, as what I am getting with mediabox, etc. is similar to the example below.

**Page size:      415.79 x 1009.81 pts**
Page rot:       0
MediaBox:           0.00     0.00   432.00  1152.00
CropBox:            7.89    16.19   423.68  1026.00
BleedBox:           7.89    16.19   423.68  1026.00
TrimBox:            7.89    16.19   423.68  1026.00
ArtBox:             7.89    16.19   423.68  1026.00

**Comment by pubpub-zz** [comment]
![image](https://github.com/py-pdf/pypdf/assets/4083478/3f46032a-b13b-42ec-93d1-68def6cc638f)

I would assume you are in the situation where the entry is missing or set to 1, then I assume the page size (mediabox) is  432x1152 = 6"x16"
the page size showed in here should be the printed area (within the crop) 415.79=423.68-7.89 x 1009.81=1026-16.19
if you want more information about the words "Page size" you should deal with pdfbox forums

**Reply by devsetgo** [reply]
Again, thank you for being patient. The PDF format is a bit obtuse and I am trying to understand. 

Below is something like the code I am using at work and with a similar result.  In the example below, I took a Word document with some text and saved as a PDF (v1.7). What I am seeing is the mediabox is saying it is 8.5 x 11 size, but what I need is the margin between the page size (8.5 x 11) and where the text starts. I thought the mediabox would be where the text begins, and I needed to get the "page size" (e.g. 8.5 x11).

Again, appreciate the help.

PDF Version %PDF-1.7
Page 0
MediaBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
CropBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
TrimBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
ArtBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
BleedBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
Unit Size: 1
Page 1
MediaBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
CropBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
TrimBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
ArtBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
BleedBox: RectangleObject([0, 0, 612, 792]) divide by 72 = 8.5 x 11.0
Unit Size: 1

```
import io
from pypdf import PdfReader

pdf_content = open("pdf_sample.pdf", "rb").read()

reader = PdfReader(io.BytesIO(pdf_content))

for page_number in range(len(reader.pages)):
    try:
        text = reader.pages[page_number].extract_text(extraction_mode="layout")
        mediabox = reader.pages[page_number].mediabox
        cropbox = reader.pages[page_number].cropbox
        trimbox = reader.pages[page_number].trimbox
        artbox = reader.pages[page_number].artbox
        bleedbox = reader.pages[page_number].bleedbox
        unit_size = reader.pages[page_number].user_unit

        print(f"Page {page_number}")
        # print(f"Text: {text}")
        print(f"MediaBox: {mediabox} divid by 72 = {mediabox[2] / 72} x {mediabox[3] / 72}")
        print(f"CropBox: {cropbox} divid by 72 = {cropbox[2] / 72} x {cropbox[3] / 72}")
        print(f"TrimBox: {trimbox} divid by 72 = {trimbox[2] / 72} x {trimbox[3] / 72}")
        print(f"ArtBox: {artbox} divid by 72 = {artbox[2] / 72} x {artbox[3] / 72}")
        print(f"BleedBox: {bleedbox} divid by 72 = {bleedbox[2] / 72} x {bleedbox[3] / 72}")
        print(f"Unit Size: {unit_size}")

    except Exception as ex:
        print(f"Error on page {page_number}: {ex}")
```

**Reply by pubpub-zz** [reply]
This means that there is no use of the crop box to create a margin, but the PDF just relies on the good will of the program to position text and drawings. If your page contains only text you could use visitor functions to locate them

**Reply by stefan6419846** [reply]
As mentioned, this sounds like something you should be able to check with a visitor - at least for texts: https://pypdf.readthedocs.io/en/stable/user/extract-text.html#using-a-visitor

**Reply by devsetgo** [reply]
Thank you for all the help. I learned a few things about how I was trying to get the margin. 

First is that I used Microsoft word to create samples and MS Word does not change the mediabox values and effectively creates it fullsize. When I ran the code against other documents created specifically for printing (booklets) those do show mediabox being different that the full size of the page. Though, if just text then it will commonly show the full page size.

So any document not created in a PDF creation tool, like MS Word, will not have a margin set as it does not set the mediabox value. Or at least that is how I interpret the information.
